packages/util-dynamodb/src/convertToAttr.ts

```diff
@@ -0,0 +1,117 @@
+import { AttributeValue } from "@aws-sdk/client-dynamodb";
+
+import { NativeAttributeValue } from "./models";
+
+/**
+ * Convert a JavaScript value to its equivalent DynamoDB AttributeValue type
+ *
+ * @param {NativeAttributeValue} data The data to convert to a DynamoDB AttributeValue
+ * @param {Object} options
+ * @param {boolean} options.convertEmptyValues Whether to automatically
+ *                      convert empty strings, blobs, and sets to `null`
+ */
+export const convertToAttr = (
+  data: NativeAttributeValue,
+  options?: {
+    convertEmptyValues: boolean;
+  }
+): AttributeValue => {
+  if (Array.isArray(data)) {
+    return { L: data.map((item) => convertToAttr(item)) };
+  } else if (data?.constructor?.name === "Set") {
+    return convertToSetAttr(data as Set<any>, options?.convertEmptyValues);
+  } else if (data?.constructor?.name === "Object") {
```

Comment:
is this code borrowed from the current (v2) DocumentClient? or is this a new implementation?

if I'm reading correctly, this will only serialize pure JS Objects to JSON.
Anyone who's using ES6 classes or other easily serializable objects would be hosed without custom premarshall?

Again, borrowing the rules from JSON serialization would be $

Reply:
> is this code borrowed from the current (v2) DocumentClient? or is this a new implementation?

This PR is new implementation in TypeScript based on the behavior of v2 DocumentClient.

> Anyone who's using ES6 classes or other easily serializable objects would be hosed without custom premarshall?

Can you provide an example?

Reply:
For sure
Here is the source, btw:
https://github.com/aws/aws-sdk-js/blob/5155fde557ecc1d166bb5040823289254bea1641/lib/dynamodb/converter.js#L27

The old DocumentClient defaulted to treating objects as map. Thus

```
class FooObj {
  constructor() {
    this.foo = 'foo';
  }
}

JSON.stringify(new FooObj())
"{"foo":"foo"}"

// new implementation needs constructor name to be Object or it errors
(new FooObj()).constructor.name
"FooObj"

// old implementation checked typeof
typeof new FooObj()
"object"
```

See https://github.com/aws/aws-sdk-js/blob/cc29728c1c4178969ebabe3bbe6b6f3159436394/lib/dynamodb/types.js#L3

Reply:
This is a miss.

However, using `JSON.stringify()` for serializing and `JSON.parse()` for deserializing is a norm in JavaScript world.
* Example StackOverflow discussion: https://stackoverflow.com/a/40201783
* MDN: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify

I created an issue to discuss this internally, and will post updates in https://github.com/aws/aws-sdk-js-v3/issues/1535

Reply:
We're not using `data instanceof Object`, as it applies to binary data as shown in code below:

```js
const data = new ArrayBuffer();
data instanceof Object;
```

Reply:
Would checking in order of `set`, `list`, `map`, `scalars` help?

Reply:
> Would checking in order of `set`, `list`, `map`, `scalars` help?

Did you mean moving check for `data instanceof Object` in `scalars` after String?

Reply:
> Would checking in order of `set`, `list`, `map`, `scalars` help?

The check for binary happens under `scalars`, so map check should be moved after it.
Currently, check for scalars is in an else block, so check for `map` will have to be moved inside scalars.

Reply:
Can you share an example where `data?.constructor?.name === "Object"` will fail for JS for maps because of prototype chain?